[netcore]Implement Environment.FailFast

See the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.environment.failfast?view=netcore-3.0) for full description.

Couple of call outs:
1.) The process should write out a crash dump before terminating.  It would be nice if we could use our crash reporter as I think it does that plus a bunch of other useful things.
2.) On windows, we should write to the event log.  I don't know what the behavior is on other platforms.

Which is what I'm saying; `_exit` exits without calling any termination handlers.

The only caveat though, is I think `_exit` is not declared in system headers, (being a libc impl detail) so you get a warning from the compiler. C99 *does* standardize `_Exit` though, which is functionally equivalent, and systems sometimes have their own implementation too (like AIX and `quick_exit` though it supports the C99 way too). Example:

```c
#include <stdio.h>
#include <stdlib.h>

void x(void) {
        printf("atexit handler reached\n");
}

int main(void){
        atexit(x);
        printf("exiting\n");
#ifdef _EXIT
        _exit(0);
#else
        exit(0);
#endif
        return 1;
}
```

```
$ gcc -maix64 -o atexit-exit atexit.c
$ gcc -maix64 -o atexit-_exit -D_EXIT atexit.c
atexit.c: In function 'main':
atexit.c:12:2: warning: implicit declaration of function '_exit' [-Wimplicit-function-declaration]
  _exit(0);
  ^~~~~
atexit.c:12:2: warning: incompatible implicit declaration of built-in function '_exit'
```

```
aix$ ./atexit-_exit
exiting
aix$ ./atexit-exit
exiting
atexit handler reached
```

--- a/mcs/class/corlib/System/Environment.cs
+++ b/mcs/class/corlib/System/Environment.cs
@@ -920,21 +920,22 @@ namespace System {
 		[SecurityPermission (SecurityAction.LinkDemand, UnmanagedCode=true)]
 		public static void FailFast (string message)
 		{
-			throw new NotImplementedException ();
+			FailFast (message, null, null);
 		}
 
 		internal static void FailFast (String message, uint exitCode)
 		{
-			throw new NotImplementedException ();
+			FailFast (message, null, null);
 		}
 
 		[SecurityCritical]
 		public static void FailFast (string message, Exception exception)
 		{
-#pragma warning disable 618
-			throw new ExecutionEngineException (message, exception);
-#pragma warning restore
+			FailFast (message, exception, null);
 		}
+
+		[MethodImplAttribute (MethodImplOptions.InternalCall)]
+		internal extern static void FailFast (string message, Exception exception, string errorSource);
 
 		[MethodImplAttribute (MethodImplOptions.InternalCall)]
 		extern static bool GetIs64BitOperatingSystem ();

--- a/mono/metadata/icall-def-netcore.h
+++ b/mono/metadata/icall-def-netcore.h
@@ -78,6 +78,7 @@ HANDLES(ENUM_5, "InternalHasFlag", ves_icall_System_Enum_InternalHasFlag, MonoBo
 
 ICALL_TYPE(ENV, "System.Environment", ENV_1)
 NOHANDLES(ICALL(ENV_1, "Exit", ves_icall_System_Environment_Exit))
+HANDLES(ENV_1a, "FailFast", ves_icall_System_Environment_FailFast, void, 3, (MonoString, MonoException, MonoString))
 HANDLES(ENV_2, "GetCommandLineArgs", ves_icall_System_Environment_GetCommandLineArgs, MonoArray, 0, ())
 ICALL(ENV_3, "GetEnvironmentVariableNames", ves_icall_System_Environment_GetEnvironmentVariableNames)
 HANDLES(ENV_8, "InternalSetEnvironmentVariable", ves_icall_System_Environment_InternalSetEnvironmentVariable, void, 4, (const_gunichar2_ptr, gint32, const_gunichar2_ptr, gint32))

--- a/mono/metadata/icall-def.h
+++ b/mono/metadata/icall-def.h
@@ -312,6 +312,7 @@ HANDLES(ENUM_7, "get_value", ves_icall_System_Enum_get_value, MonoObject, 1, (Mo
 
 ICALL_TYPE(ENV, "System.Environment", ENV_1)
 NOHANDLES(ICALL(ENV_1, "Exit", ves_icall_System_Environment_Exit))
+HANDLES(ENV_1a, "FailFast", ves_icall_System_Environment_FailFast, void, 3, (MonoString, MonoException, MonoString))
 HANDLES(ENV_2, "GetCommandLineArgs", ves_icall_System_Environment_GetCommandLineArgs, MonoArray, 0, ())
 ICALL(ENV_3, "GetEnvironmentVariableNames", ves_icall_System_Environment_GetEnvironmentVariableNames)
 NOHANDLES(ICALL(ENV_31, "GetIs64BitOperatingSystem", ves_icall_System_Environment_GetIs64BitOperatingSystem))

--- a/mono/metadata/icall.c
+++ b/mono/metadata/icall.c
@@ -7410,7 +7410,7 @@ ves_icall_System_Environment_FailFast (MonoStringHandle message, MonoExceptionHa
 	// information in the error dump that CoreCLR would. On Windows 7+ we should call
 	// RaiseFailFastException directly instead of relying on the C runtime doing it
 	// for us and pass it as much information as possible. On Windows 8+ we can also
-	// use the __failfast intrinsic.
+	// use the __fastfail intrinsic.
 	abort ();
 }
 

--- a/mono/metadata/icall.c
+++ b/mono/metadata/icall.c
@@ -7409,7 +7409,8 @@ ves_icall_System_Environment_FailFast (MonoStringHandle message, MonoExceptionHa
 	// NOTE: While this does trigger WER on Windows it doesn't quite provide all the
 	// information in the error dump that CoreCLR would. On Windows 7+ we should call
 	// RaiseFailFastException directly instead of relying on the C runtime doing it
-	// for us and pass it as much information as possible.
+	// for us and pass it as much information as possible. On Windows 8+ we can also
+	// use the __failfast intrinsic.
 	abort ();
 }
 

--- a/mono/metadata/icall.c
+++ b/mono/metadata/icall.c
@@ -7391,6 +7391,28 @@ ves_icall_System_Environment_Exit (int result)
 	exit (result);
 }
 
+void
+ves_icall_System_Environment_FailFast (MonoStringHandle message, MonoExceptionHandle exception, MonoStringHandle errorSource, MonoError *error)
+{
+	if (MONO_HANDLE_IS_NULL (message)) {
+		g_warning ("CLR: Managed code called FailFast without specifying a reason.");
+	} else {
+		char *msg = mono_string_handle_to_utf8 (message, error);
+		g_warning ("CLR: Managed code called FailFast, saying \"%s\"", msg);
+		g_free (msg);
+	}
+
+	if (!MONO_HANDLE_IS_NULL (exception)) {
+		mono_print_unhandled_exception_internal ((MonoObject *) MONO_HANDLE_RAW (exception));
+	}
+
+	// NOTE: While this does trigger WER on Windows it doesn't quite provide all the
+	// information in the error dump that CoreCLR would. On Windows 7+ we should call
+	// RaiseFailFastException directly instead of relying on the C runtime doing it
+	// for us and pass it as much information as possible.
+	abort ();
+}
+
 MonoStringHandle
 ves_icall_System_Environment_GetGacPath (MonoError *error)
 {

--- a/netcore/CoreFX.issues.rsp
+++ b/netcore/CoreFX.issues.rsp
@@ -175,16 +175,6 @@
 # https://github.com/mono/mono/blob/18f0bf055eb2bfa8981062bdad3af2fb4bbc9be4/netcore/System.Private.CoreLib/shared/System/String.cs#L811-L821
 -nomethod System.Tests.StringTests.NormalizationTest
 
-# Environment.FailFast is not implemented
--nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterDouble_C20
--nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterSingle_C20
--nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterSingle_N20
--nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterSingle_P20
--nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterDouble_N20
--nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterDouble_F20
--nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterDouble_P20
--nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterSingle_F20
-
 -nomethod System.SpanTests.SpanTests.ReadOnlyMemory_PropertyReturningReadOnlySpan
 -nomethod System.SpanTests.SpanTests.Span_StaticOperator
 -nomethod System.SpanTests.SpanTests.ReadOnlySpan_Constructor
@@ -801,11 +791,6 @@
 ####################################################################
 ##  System.Runtime.Extensions.Tests
 ####################################################################
-
-# Environment.FailFast is not implemented
--nomethod System.Tests.EnvironmentTests.FailFast_ExceptionStackTrace_InnerException
--nomethod System.Tests.EnvironmentTests.FailFast_ExceptionStackTrace_ArgumentException
--nomethod System.Tests.EnvironmentTests.FailFast_ExceptionStackTrace_StackOverflowException
 
 # GC.GetGCMemoryInfo is not implemented
 -nomethod System.Tests.AppDomainTests.MonitoringIsEnabled

--- a/netcore/CoreFX.issues.rsp
+++ b/netcore/CoreFX.issues.rsp
@@ -175,6 +175,16 @@
 # https://github.com/mono/mono/blob/18f0bf055eb2bfa8981062bdad3af2fb4bbc9be4/netcore/System.Private.CoreLib/shared/System/String.cs#L811-L821
 -nomethod System.Tests.StringTests.NormalizationTest
 
+# Hits Debug.Assert in debug configuration for floating point precision on round-tripping
+-nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterDouble_C20
+-nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterSingle_C20
+-nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterSingle_N20
+-nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterSingle_P20
+-nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterDouble_N20
+-nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterDouble_F20
+-nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterDouble_P20
+-nomethod System.Buffers.Text.Tests.RealFormatterTests.TestFormatterSingle_F20
+
 -nomethod System.SpanTests.SpanTests.ReadOnlyMemory_PropertyReturningReadOnlySpan
 -nomethod System.SpanTests.SpanTests.Span_StaticOperator
 -nomethod System.SpanTests.SpanTests.ReadOnlySpan_Constructor

--- a/netcore/System.Private.CoreLib/src/System/Environment.cs
+++ b/netcore/System.Private.CoreLib/src/System/Environment.cs
@@ -54,18 +54,16 @@ namespace System
 
 		public static void FailFast (string message)
 		{
-			throw new NotImplementedException ();
+			FailFast (message, null, null);
 		}
 
 		public static void FailFast(string message, Exception exception)
 		{
-			throw new NotImplementedException ();
+			FailFast (message, exception, null);
 		}
 
-		public static void FailFast (string message, Exception exception, string errorMessage)
-		{
-			throw new NotImplementedException ();
-		}
+		[MethodImplAttribute (MethodImplOptions.InternalCall)]
+		public extern static void FailFast (string message, Exception exception, string errorSource);
 	}
 
 #region referencesource dependencies - to be removed

